pkg/finality-grandpa/voting_round.go

```diff
@@ -0,0 +1,849 @@
+				primary := PrimaryPropose[Hash, Number]{
+					TargetHash:   lastRoundEstimate.Hash,
+					TargetNumber: lastRoundEstimate.Number,
+				}
+				err := vr.env.Proposed(vr.roundNumber(), primary)
+				if err != nil {
+					return err
+				}
+				message := newMessage(primary)
+				vr.outgoing.Push(message)
+				vr.state = stateProposed[Timer]{prevoteTimer, precommitTimer}
+
+				return nil
+			}
+			fmt.Printf(
+				"Last round estimate has been finalized, not sending primary block hint for round %d\n",
+				vr.votes.Number(),
+			)
+
+		case maybeEstimate == nil && vr.voting.isPrimary():
+			fmt.Printf("Last round estimate does not exist, not sending primary block hint for round %d\n", vr.votes.Number())
+		default:
+		}
+
+		vr.state = stateStart[Timer]{prevoteTimer, precommitTimer}
+	default:
+		vr.state = state
+	}
+	return nil
+}
+
+func (vr *votingRound[Hash, Number, Signature, ID, E]) prevote(w *waker, lastRoundState *RoundState[Hash, Number]) error { //nolint:lll
+	state := vr.state
+	vr.state = nil
+
+	var startPrevoting = func(prevoteTimer Timer, precommitTimer Timer, proposed bool, waker *waker) error {
+		prevoteTimer.SetWaker(waker)
+		var shouldPrevote bool
+		elapsed, err := prevoteTimer.Elapsed()
+		if elapsed {
+			if err != nil {
+				return err
+			}
+			shouldPrevote = true
+		} else {
+			shouldPrevote = vr.votes.Completable()
+		}
+
+		if shouldPrevote {
+			if vr.voting.isActive() {
+				fmt.Println("Constructing prevote for round", vr.votes.Number())
+
+				base, bestChain := vr.constructPrevote(lastRoundState)
+
+				// since we haven't polled the future above yet we need to
+				// manually schedule the current task to be awoken so the
+				// `best_chain` future is then polled below after we switch the
+				// state to `Prevoting`.
+				waker.wake()
+
+				vr.state = statePrevoting[Timer, hashBestChain[Hash, Number]]{
+					precommitTimer, hashBestChain[Hash, Number]{base, bestChain},
+				}
+			} else {
+				vr.state = statePrevoted[Timer]{precommitTimer}
+			}
+		} else if proposed {
+			vr.state = stateProposed[Timer]{prevoteTimer, precommitTimer}
+		} else {
+			vr.state = stateStart[Timer]{prevoteTimer, precommitTimer}
+		}
+
+		return nil
+	}
+
+	var finishPrevoting = func(precommitTimer Timer, base Hash, bestChain BestChain[Hash, Number], waker *waker) error {
+		wakerChan := newWakerChan(bestChain)
+		wakerChan.setWaker(waker)
+		var best *HashNumber[Hash, Number]
+		res := <-wakerChan.channel()
+		switch {
+		case res.Error != nil:
+			return res.Error
+		case res.Value != nil:
+			best = res.Value
+		default:
+			vr.state = statePrevoting[Timer, hashBestChain[Hash, Number]]{
+				precommitTimer, hashBestChain[Hash, Number]{base, bestChain},
+			}
+			return nil
+		}
+
+		if best != nil {
+			prevote := Prevote[Hash, Number]{best.Hash, best.Number}
+
+			// TODO: debug!(target: "afg", "Casting prevote for round {}", this.votes.number());
```

Comment:
nit/remove log comments in this function

Reply:
I need to figure out the logging aspect of this, then I'll remove the TODOs.